src/digodoc_lib/cmt.ml

```diff
@@ -64,12 +64,192 @@ let pp_val_desc ident pp vdesc =
+                                Format.pp_print_string fmt s
+                            )) l
+            | Otyp_module (id, _s_l, _out_ty_l) -> !Oprint.out_ident fmt id
+            | Otyp_class (_b, out_id, _out_type_l) ->
+                    Format.fprintf fmt "%a"!Oprint.out_ident out_id
+            | Otyp_constr (out_id, _out_type_l) ->
+                        Format.fprintf fmt "%a"!Oprint.out_ident out_id
+
+            | Otyp_tuple _ | Otyp_var _ | Otyp_arrow _ | Otyp_attribute _ -> ()
+    in
+    let ty = Format.asprintf "%a" aux_out_type sign in
+    Format.fprintf fmt "%s" (standardise_output ty)
+
+    and pp_outvar fmt =
+    let open Outcometree in
+    function
+        | Ovar_fields list_of_fields ->
+                Format.fprintf fmt "%a" (Format.pp_print_list
+                                            ~pp_sep:(fun fmt () -> Format.fprintf fmt "@.")
+                                            (fun fmt (name, _b, _out_ty_l) ->
+                                                Format.fprintf fmt "`%s" name
+                                            )) list_of_fields
+        | Ovar_typ out_ty -> Format.fprintf fmt "`%a" out_type out_ty
+
+(** [pp_type fmt (acc, tree)] takes a type declaration tree and prints its content with function [out_type]*)
+let rec pp_type fmt (acc, tree) =
+    let open Outcometree in
+    (*
+        For documentation on the type used for representing type declarations see:
+           https://docs.ocaml.pro/docs/LIBRARY.ocamlcommon@ocaml-base-compiler.4.10.0/Outcometree/index.html#type-out_sig_item
+     *)
+    match tree with
+    | [] -> Format.fprintf fmt "%a" (Format.pp_print_list
+                                            ~pp_sep:(fun fmt () -> Format.fprintf fmt "")
+                                            (fun fmt ty -> Format.fprintf fmt "%s" ty)
+                                    ) (List.rev acc)
+    | Osig_type (out_type_decl, _out_rec_status) :: r ->
+            let ty =
+                Format.asprintf "%a" out_type out_type_decl.otype_type
+            in
+            pp_type fmt (standardise_output ty :: acc, r)
+    | _ -> failwith "should not occur"
+
+(** [pp_constructors fmt const_decl_l] Prints the constructors composing a Variant type *)
+let pp_constructors fmt const_decl_l =
+    let constructors =
+        Format.asprintf "%a"
+            (Format.pp_print_list
+                    ~pp_sep:(fun fmt ()-> Format.fprintf fmt "@.")
+                    (fun fmt constr ->
+                        Format.fprintf fmt "%s" (infix (Ident.name constr.cd_id))
+                    )
+            ) const_decl_l
+            in
+            Format.fprintf fmt "%s" (standardise_output constructors)
+
+
+(** [pp_class_sig_item fmt sig_item] Prints the items of a class *)
+let pp_class_sig_item fmt sig_item =
+    let open Outcometree in
+    match sig_item with
+    | Ocsg_constraint ( _out_ty1, _out_ty2 ) ->
+            Format.fprintf fmt "constraint"
```

Comment:
I believe this would break listing. Since it's guaranteed that a constraint will be found in amongst a `CLASS_OBJ` declaration, printing an empty line would be a problem when constructors are gathered in `digodoc-index-api`

Reply:
Why this should be a problem? You could choose what will you list and what not. Since you have at least your two first line with class name and “CLASS_OBJ” it won’t break anything (even if it breaks, it’s means that it is problem of algorithm of listing). And you are not forced, to print an empty line, just don’t print anything (even new line character). By the way, since you print always “constraint”, that means that when in the same class you ll have two constraints, that will for sure break a listing.

Reply:
Solution is to not print new line character inside ~pp_sep at the line 200. Just print it with “val ..@.” and “method ..@.”.